### PR TITLE
Al/BE-96 Hypothetical liquidity.

### PR DESCRIPTION
**Pull request by lut1k**
1. Added a oracle pallet to runtime.
Implement function for getting the underlying price of a mToken asset: get_underlying_price();
params: mToken: CurrencyId - the mToken to get the underlying price of
return: price - the underlying asset price;
Freeze return value.

2. Added get_hypothetical_account_liquidity function to controller pallet.

3. Added redeem, borrow and repay allowed functions to controller pallet.

4. Added const PoolAccountId to module in liquidity-pools pallet.

### CODE REVIEW

File: pallets/controller/src/lib.rs [hunk 1]
```diff
@@ -331,17 +341,159 @@ impl<T: Trait> Module<T> {
...
+				continue;
+			}
+
+			// Read the balances and exchange rate from the cToken
+			let borrow_balance = Self::borrow_balance_stored(account, underlying_asset)?;
+			let exchange_rate = Self::get_exchange_rate(underlying_asset)?;
+			let collateral_factor = Self::get_collateral_factor(underlying_asset);
+
+			// Get the normalized price of the asset.
+			let oracle_price =
```
Comment (Apollo50) [inline pallets/controller/src/lib.rs:376]: After getting oracle price you have to add zero checking.

File: pallets/controller/src/lib.rs [hunk 2]
```diff
@@ -331,17 +341,159 @@ impl<T: Trait> Module<T> {
...
+		for asset in m_tokens_ids.into_iter() {
+			let underlying_asset = Self::get_underlying_asset_id_by_wrapped_id(&asset)?;
+			let m_token_balance = T::MultiCurrency::free_balance(asset, account);
+			if m_token_balance == Balance::zero() {
+				continue;
+			}
+
+			// Read the balances and exchange rate from the cToken
+			let borrow_balance = Self::borrow_balance_stored(account, underlying_asset)?;
+			let exchange_rate = Self::get_exchange_rate(underlying_asset)?;
```
Comment (Apollo50) [inline pallets/controller/src/lib.rs:372]: You are getting wrong exchange rate for iteration for asset = cDots rate equals 1.4, and in iteration for assets = cETH equals 0.75 (instead 1 ). Check the reason.

File: pallets/controller/src/lib.rs [hunk 3]
```diff
@@ -535,7 +687,24 @@ impl<T: Trait> Module<T> {
 		Ok(new_borrow_index)
 	}
 
-	fn get_wrapped_id_by_underlying_asset_id(asset_id: &CurrencyId) -> CurrencyIdResult {
+	/// Returning: value += balance_scalar * rate_scalar
+	fn mul_price_and_balance_add_to_prev_value(
+		value: Balance,
```
Comment (Apollo50) [inline pallets/controller/src/lib.rs:692]: If you use this function only for getHypotetical..., better to rename params as in a main function, imho.

File: pallets/controller/src/mock.rs [hunk 4]
```diff
@@ -261,6 +246,71 @@ impl ExtBuilder {
...
+		]);
+		self.pools.push((
+			CurrencyId::DOT,
+			Pool {
+				current_interest_rate: Rate::from_inner(0),
+				total_borrowed: Balance::zero(),
+				borrow_index: Rate::saturating_from_rational(1, 1),
+				current_exchange_rate: Rate::saturating_from_rational(8, 10),
+				is_lock: false,
+				total_insurance: 5,
```
Comment (Apollo50) [inline pallets/controller/src/mock.rs:263]: For my view this params is a reason you are getting wrong exchange rate.

**Review by Apollo50 — APPROVED** [review]

File: pallets/controller/src/mock.rs [hunk 5]
```diff
@@ -261,6 +246,71 @@ impl ExtBuilder {
...
+			},
+		));
+		self.pool_user_data.push((
+			ALICE,
+			CurrencyId::DOT,
+			PoolUserData {
+				total_borrowed: 30,
+				interest_index: Rate::saturating_from_rational(2, 1),
```
Comment (Apollo50) [inline pallets/controller/src/mock.rs:306]: for start interest_index has to be equal 1

**Review by Apollo50 — APPROVED** [review]

File: pallets/controller/src/lib.rs (hunk 1 above)
Comment (lut1k) [reply pallets/controller/src/lib.rs:376]: Ok, I made this change on my new branch AL/BE-99 ....

File: pallets/controller/src/lib.rs (hunk 3 above)
Comment (lut1k) [reply pallets/controller/src/lib.rs:692]: We discussed it. Decided to leave unchanged.

File: pallets/controller/src/mock.rs (hunk 4 above)
Comment (lut1k) [reply pallets/controller/src/mock.rs:263]: Fixed.

File: pallets/controller/src/mock.rs (hunk 5 above)
Comment (lut1k) [reply pallets/controller/src/mock.rs:306]: Fixed.

File: pallets/controller/src/lib.rs [hunk 6]
```diff
@@ -331,17 +341,159 @@ impl<T: Trait> Module<T> {
...
+			// sum_collateral += tokens_to_denom * m_token_balance
+			sum_collateral =
+				Self::mul_price_and_balance_add_to_prev_value(sum_collateral, m_token_balance, tokens_to_denom)?;
+
+			// sum_borrow_plus_effects += oracle_price * borrow_balance
+			sum_borrow_plus_effects =
+				Self::mul_price_and_balance_add_to_prev_value(sum_borrow_plus_effects, borrow_balance, oracle_price)?;
+
+			// Calculate effects of interacting with Underlying Asset Modify
+			if underlying_asset_id == underlying_asset {
```
Comment (Pavel-Blaize) [inline pallets/controller/src/lib.rs:395]: just a bit confusing "underlying_asset_id" vs "underlying_asset"
may be "underlying_asset_id" -> "underlying_to_borrow" or something like that?

File: pallets/controller/src/lib.rs [hunk 7]
```diff
@@ -331,17 +341,159 @@ impl<T: Trait> Module<T> {
...
+				// redeem effect
+				// sum_borrow_plus_effects += tokens_to_denom * redeem_tokens
+				sum_borrow_plus_effects = Self::mul_price_and_balance_add_to_prev_value(
```
Comment (Pavel-Blaize) [inline pallets/controller/src/lib.rs:398]: in a single time 1 or None of these 2 action are performed. Probably we need to add checks:
if (redeem_amount > 0) { increase sum_borrow_plus_effects with redeem_amoont; }
if (borrow_amount > 0) { increase sum_borrow_plus_effects with redeem_amoont; }

File: pallets/controller/src/lib.rs (hunk 6 above)
Comment (lut1k) [reply pallets/controller/src/lib.rs:395]: Fixed. Made changes in a new branch.

File: pallets/controller/src/lib.rs [hunk 8]
```diff
@@ -331,17 +341,159 @@ impl<T: Trait> Module<T> {
...
-		let _borrow_amount = borrow_amount;
-		Ok((Balance::zero(), Balance::zero()))
+		let m_tokens_ids = T::MTokensId::get();
+
+		let mut sum_collateral = Balance::zero();
+		let mut sum_borrow_plus_effects = Balance::zero();
+
+		// For each tokens the account is in
+		for asset in m_tokens_ids.into_iter() {
+			let underlying_asset = Self::get_underlying_asset_id_by_wrapped_id(&asset)?;
```
Comment (Pavel-Blaize) [inline pallets/controller/src/lib.rs:364]: Check for check_user_available_collateral() should be added, so we will skip calculations for the m-tokens user doesn't want to use as a collateral

**Review by Pavel-Blaize — APPROVED** [review]

File: pallets/controller/src/lib.rs (hunk 7 above)
Comment (lut1k) [reply pallets/controller/src/lib.rs:398]: Fixed. Made changes in a new branch.

File: pallets/controller/src/lib.rs (hunk 8 above)
Comment (lut1k) [reply pallets/controller/src/lib.rs:364]: Fixed. Made changes in a new branch.

**Pull request merged**